jobs/mongodb_backup_job.go

```diff
@@ -0,0 +1,261 @@
+	}
+
+	if err := j.waitUntilBackupCompletion(ctx); err != nil {
+		return errors.Wrap(err, "failed to wait backup completion")
+	}
+
+	send(&agentpb.JobResult{
+		JobId:     j.id,
+		Timestamp: ptypes.TimestampNow(),
+		Result: &agentpb.JobResult_MongodbBackup{
+			MongodbBackup: &agentpb.JobResult_MongoDBBackup{},
+		},
+	})
+
+	return nil
+}
+
+func createDBURL(dbConfig DatabaseConfig) *url.URL {
+	var host string
+	switch {
+	case dbConfig.Address != "":
+		if dbConfig.Port > 0 {
+			host = net.JoinHostPort(dbConfig.Address, strconv.Itoa(dbConfig.Port))
+		} else {
+			host = dbConfig.Address
+		}
+	case dbConfig.Socket != "":
+		host = dbConfig.Socket
+	}
+
+	var user *url.Userinfo
+	if dbConfig.User != "" {
+		user = url.UserPassword(dbConfig.User, dbConfig.Password)
+	}
+
+	return &url.URL{
+		Scheme: "mongodb",
+		User:   user,
+		Host:   host,
+	}
+}
+
+func (j *MongoDBBackupJob) startBackup(ctx context.Context) error {
+	j.l.Info("Starting backup.")
+
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	output, err := exec.CommandContext(nCtx, pbmBin, "backup", "--mongodb-uri="+j.dbURL.String()).CombinedOutput() // #nosec G204
+
+	if err != nil {
+		return errors.Wrapf(err, "pbm backup error: %s", string(output))
+	}
+
+	return nil
+}
+
+func (j *MongoDBBackupJob) checkBackupCompletion(ctx context.Context) (bool, error) {
+	j.l.Debug("Checking backup status.")
+
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	output, err := exec.CommandContext(nCtx, pbmBin, "status", "--mongodb-uri="+j.dbURL.String()).CombinedOutput() // #nosec G204
+	if err != nil {
+		return false, errors.Wrapf(err, "pbm status error: %s", string(output))
+	}
+
+	return backupStatusOutputR.Match(output), nil
+}
+
+func (j *MongoDBBackupJob) waitUntilBackupCompletion(ctx context.Context) error {
+	j.l.Info("Waiting for backup completion.")
+	ticker := time.NewTicker(backupStatusCheckInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			done, err := j.checkBackupCompletion(ctx)
+			if err != nil {
+				return errors.Wrapf(err, "failed to check backup completion")
+			}
+
+			if done {
+				return nil
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (j *MongoDBBackupJob) setupS3(ctx context.Context) error {
+	j.l.Info("Configuring S3 location.")
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	confFile, err := j.writePBMConfigFile()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer os.Remove(confFile) //nolint:errcheck
+
+	output, err := exec.CommandContext( //nolint:gosec
+		nCtx,
+		pbmBin,
+		"config",
+		"--mongodb-uri="+j.dbURL.String(),
+		"--file="+confFile,
+	).CombinedOutput()
+
+	if err != nil {
+		return errors.Wrapf(err, "pbm config error: %s", string(output))
+	}
+
+	return nil
+}
+
+func (j *MongoDBBackupJob) writePBMConfigFile() (string, error) {
+	tmp, err := ioutil.TempFile("", "pbm-config-*.yml")
```

Comment:
can we defer Close here, so you don't need to close it each time when error occured?

Reply:
https://www.joeshaw.org/dont-defer-close-on-writable-files/

We need an error from `tmp.Close` when it's the last step, because it can mean that file is corrupted, but we need just to try to close a file in case of other errors.